src/xenia/cpu/xex_module.cc

```diff
@@ -178,57 +707,387 @@ bool XexModule::ApplyPatch(XexModule* module) {
+
+// The macros in pe_image.h don't work with clang, for some reason.
+// offsetof seems to be unable to find OptionalHeader.
+#define offsetof1(type, member) ((std::size_t) & (((type*)0)->member))
+#define IMAGE_FIRST_SECTION1(ntheader)                                   \
+  ((PIMAGE_SECTION_HEADER)(                                              \
+      (uint8_t*)ntheader + offsetof1(IMAGE_NT_HEADERS, OptionalHeader) + \
+      ((PIMAGE_NT_HEADERS)(ntheader))->FileHeader.SizeOfOptionalHeader))
+
+  // Quick scan to determine bounds of sections.
+  size_t upper_address = 0;
+  const IMAGE_SECTION_HEADER* sechdr = IMAGE_FIRST_SECTION1(nthdr);
+  for (size_t n = 0; n < filehdr->NumberOfSections; n++, sechdr++) {
+    const size_t physical_address = opthdr->ImageBase + sechdr->VirtualAddress;
+    upper_address =
+        std::max(upper_address, physical_address + sechdr->Misc.VirtualSize);
+  }
+
+  // Setup/load sections.
+  sechdr = IMAGE_FIRST_SECTION1(nthdr);
+  for (size_t n = 0; n < filehdr->NumberOfSections; n++, sechdr++) {
+    PESection section;
+    memcpy(section.name, sechdr->Name, sizeof(sechdr->Name));
+    section.name[8] = 0;
+    section.raw_address = sechdr->PointerToRawData;
+    section.raw_size = sechdr->SizeOfRawData;
+    section.address = base_address_ + sechdr->VirtualAddress;
+    section.size = sechdr->Misc.VirtualSize;
+    section.flags = sechdr->Characteristics;
+    pe_sections_.push_back(section);
+    // pe_sections_.push_back(section);
+  }
+
+  // DumpTLSDirectory(pImageBase, pNTHeader, (PIMAGE_TLS_DIRECTORY32)0);
+  // DumpExportsSection(pImageBase, pNTHeader);
+  return 0;
 }
 
 bool XexModule::Load(const std::string& name, const std::string& path,
-                     xe_xex2_ref xex) {
+                     const void* xex_addr, size_t xex_length,
+                     XexModule* base_xex) {
+  auto src_header = reinterpret_cast<const xex2_header*>(xex_addr);
+
+  if (src_header->magic != 'XEX2') {
+    return false;
+  }
+
   assert_false(loaded_);
   loaded_ = true;
-  xex_ = xex;
 
-  auto old_header = xe_xex2_get_header(xex_);
+  base_xex_ = base_xex;
+
+  // Read in XEX headers
+  xex_header_mem_.resize(src_header->header_size);
+  std::memcpy(xex_header_mem_.data(), src_header, src_header->header_size);
+
+  auto sec_header = xex_security_info();
+
+  // Try setting our base_address based on XEX_HEADER_IMAGE_BASE_ADDRESS, fall
+  // back to xex_security_info otherwise
+  base_address_ = xex_security_info()->load_address;
+  xe::be<uint32_t>* base_addr_opt = nullptr;
+  if (GetOptHeader(XEX_HEADER_IMAGE_BASE_ADDRESS, &base_addr_opt))
+    base_address_ = *base_addr_opt;
 
   // Setup debug info.
   name_ = std::string(name);
   path_ = std::string(path);
-  // TODO(benvanik): debug info
+
+  uint8_t* data = memory()->TranslateVirtual(base_address_);
+
+  // Load in the XEX basefile
+  // We'll try using both XEX2 keys to see if any give a valid PE
+  while (true) {
```

Comment:
Instead of doing a loop here to try both decryption keys, it'd be better to just try loading the xex with the retail key first, and if that fails, try with the devkit key.
If they both fail, error out.

Reply:
Isn't --xex_dev_key used to tell xenia to use the dev key?

Reply:
As in passing something like "bool use_dev_key" to XexModule? That probably would be better.

@Razzile this PR tries both keys right now, so the xex_dev_key option is removed